### ISSUE TITLE
React App crashes with "Cannot use 'import.meta' outside a module" after updating react-router-dom

### ISSUE THREAD

**Issue by gonzadelarge** [meteor/meteor#13504]
After updating _react-router-dom_ from version **6.28.0** to **7.0.1**, my app crashes, and the following error appears in the browser console:

**Uncaught SyntaxError: Cannot use 'import.meta' outside a module**

I am using **METEOR@3.0.4**, React for the front-end, and _react-router-dom_ for routing. The error occurs only after upgrading _react-router-dom_. The app works as expected with version **6.28.0**.

**Steps to Reproduce**
1 - meteor create app --react
2 -  meteor npm i react-router-dom
3 - start the app an check the browser console

**Environment**
Meteor 3.0.4
Node v20.17.0
React 18.2.0
React-router-dom : 7.0.1

The error message suggests the use of **import.meta**, which is incompatible outside ES modules. This issue appears to be related to how Meteor handles modules. The problem does not occur with r### eact-router-dom@6.28.0

**Comment by nachocodoner** [comment]
https://forums.meteor.com/t/react-router-v7-not-working-with-meteor

There’s some discussion on the forums about this issue, including a workaround involving changes to how modules are exported in the new React Router version, with possibly to lock this changes with [patch-package](https://www.npmjs.com/package/patch-package).

The problem could be from a limitation in the Meteor bundler (pending debugging for confirmation), which doesn’t handle the library’s new .mjs files as expected. We will research on possibilities to fully solve this, but the modern bundler integration could be the best to get rid of this inconveniences coming over time.

**Comment by gvanhorne** [comment]
Jumping in to +1 this issue and outline the steps I took as a workaround following paulishca's solution here: https://forums.meteor.com/t/react-router-v7-not-working-with-meteor/62618/6

1. Add `"postinstall": "patch-package"` to your project's `package.json` scripts
2. Install [patch-package](https://www.npmjs.com/package/patch-package) with `npm i patch-package` 
3. replace all `.mjs` extensions with `.js` in `node_modules/react-router/package.json`
4. Apply the patch with `npx patch-package react-router --exclude`

The patch is now applied automatically after installing packages.

**Comment by 9Morello** [comment]
As an alternative workaround, using the latest version of the [Vite bundler package](https://github.com/JorgenVatle/meteor-vite) from @JorgenVatle makes this package work without issues.

**Comment by kakadais** [comment]
Alternative workaround is installing 
```
meteor npm install react-router-dom@6
```
It causes some future flag warning but seems like working for now.

Any update for this issue?

**Comment by nachocodoner** [comment]
I researched this by debugging the Meteor core and found that the issue is related to the support of export fields in the Meteor bundler. react-router 7 uses exports field on package.json. It seems connected to this PR: https://github.com/meteor/meteor/pull/13520.

I'll assist with testing to the new PR with exports fields to check if react-router may be fixed by it and ensure it's published soon in an upcoming release.

**Comment by domq** [comment]
This issue is absolutely current. @github-actions please bump.

**Comment by JorgenVatle** [comment]
Somewhat related, but not really addressing the main issue - There's a hacky workaround to the limitation where the Meteor server will crash on startup when `"type": "module"` is set in `package.json`.

The internal Meteor server entrypoint, the one that's executed when you run `meteor run` in the terminal, uses `require()` instead of `import`, unless I'm mistaken. Which by default is within `.meteor/local` relative to your app's root directory.

So if you symlink Meteor's local directory outside of your application root, CommonJS becomes the default again. For example:
```shell
mkdir -p /tmp/.meteor-local/my-app
ln -s /tmp/.meteor-local/my-app .meteor/local
```

Then if you pre-bundle your application code, you can get around the `import.meta` limitation. E.g. using something like [`tsup --watch`](https://github.com/egoist/tsup) and use the build output as the new mainModule for your Meteor server/client.

Though, a limitation of this is that imports for Meteor packages may have to be stubbed out or externalized for `tsup` to not yell at you.

**Comment by antonio-goncalves** [comment]
Get this error when running the postinstall script, anyone having the same issue?

patch-package 8.0.0
Applying patches...

**ERROR** Failed to apply patch for package react-router at path
  
  node_modules/react-router

  This error was caused because patch-package cannot apply the following patch file:

  patches/react-router+7.9.2.patch

  Try removing node_modules and trying again. If that doesn't work, maybe there was
  an accidental change made to the patch file? Try recreating it by manually
  editing the appropriate files and running:
  
  patch-package react-router
  
  If that doesn't work, then it's a bug in patch-package, so please submit a bug
  report. Thanks!

  https://github.com/ds300/patch-package/issues

**Comment by nachocodoner** [comment]
You could re-apply the patch within node_modules/react-router, and extract the patch instructions `patch-package react-router`, as the message suggested to you. Have you tried that?

Anyway, react-router will be compatible with upcoming Meteor 3.4 and Rspack integration, [betas running already](https://forums.meteor.com/t/new-3-4-beta-11-release-faster-builds-smaller-bundles-and-modern-setups-with-the-rspack-integration).

**Comment by antonio-goncalves** [comment]
Hi,

It seems that the problem is the fact of running a old version of meteor (2.6.1)

I tried on another meteor project with the version 3.0.2 and it worked fine

**Comment by antonio-goncalves** [comment]
I just read about version 3.4 I am looking forward 💪

**Comment by nachocodoner** [comment]
Meteor 3.4 is already released. [With Rspack support](https://docs.meteor.com/about/modern-build-stack/rspack-bundler-integration.html), it gives compatibility when using React Router v7, and other ESM-only packages or packages that require a modern bundler.

Related: https://github.com/meteor/meteor/issues/12324#issuecomment-3854235081
